rackcity/utils/query_utils.py

```diff
@@ -0,0 +1,296 @@
+                if not range_serializer.is_valid():
+                    raise Exception(
+                        "Invalid rack_range filter: " +
+                        str(range_serializer.errors)
+                    )
+                filter_args.append(
+                    {
+                        'rack__rack_num__range':
+                        range_serializer.get_number_range()
+                    }
+                )
+                filter_args.append(
+                    {
+                        'rack__row_letter__range':
+                        range_serializer.get_row_range()  # noqa inclusive on both letter, number
+                    }
+                )
+
+            else:
+                raise Exception(
+                    "String field 'filter_type' must be either 'text', " +
+                    "'numeric', or 'rack_range'."
+                )
+
+    return filter_args
+
+
+def get_filtered_query(object_query, data):
+    try:
+        filter_args = get_filter_arguments(data)
+    except Exception as error:
+        return (
+            None,
+            JsonResponse(
+                {
+                    "failure_message":
+                    Status.ERROR.value + GenericFailure.FILTER.value,
+                    "errors": str(error)
+                },
+                status=HTTPStatus.BAD_REQUEST
+            )
+        )
+    for filter_arg in filter_args:
+        object_query = object_query.filter(**filter_arg)
+    return (object_query, None)
+
+
+def get_sorted_query(object_query, data):
+    try:
+        sort_args = get_sort_arguments(data)
+    except Exception as error:
+        return (
+            None,
+            JsonResponse(
+                {
+                    "failure_message":
+                    Status.ERROR.value + GenericFailure.SORT.value,
+                    "errors": str(error)
+                },
+                status=HTTPStatus.BAD_REQUEST
+            )
+        )
+    objects = object_query.order_by(*sort_args)
+    return (objects, None)
+
+
+def get_paginated_query(objects, request):
+    paginator = PageNumberPagination()
+    paginator.page_size = request.query_params.get('page_size')
+    try:
+        page_of_objects = paginator.paginate_queryset(objects, request)
+    except Exception as error:
+        return (
+            None,
+            JsonResponse(
+                {
+                    "failure_message":
+                    Status.ERROR.value + GenericFailure.PAGE_ERROR.value,
+                    "errors": str(error)
+                },
+                status=HTTPStatus.BAD_REQUEST,
+            )
+        )
+    return (page_of_objects, None)
+
+
+def should_paginate_query(query_params):
+    return not(
+        query_params.get('page') is None
+        and query_params.get('page_size') is None
+    )
+
+
+def get_many_response(model, model_serializer, response_key, request):
```

Comment:
Refactored. Returns response for commonly use "many" function view. Applies sort, filter, pagination to the requested Django model, then serializes the objects, and returns JsonResponse (fail or success).